libraries/Microsoft.Bot.Builder.LUIS/LuisRecognizer.cs

```diff
@@ -0,0 +1,217 @@
+                    JArray.FromObject(entity.Resolution["values"]);
+        }
+
+        private static JObject GetEntityMetadata(EntityRecommendation entity)
+        {
+            return JObject.FromObject(new
+            {
+                startIndex = entity.StartIndex,
+                endIndex = entity.EndIndex,
+                text = entity.Entity,
+                score = entity.Score
+            });
+        }
+
+        private static string GetNormalizedEntityType(EntityRecommendation entity)
+        {
+            return Regex.Replace(entity.Type, "\\.", "_");
+        }
+
+        private static IList<EntityRecommendation> PopulateCompositeEntity(CompositeEntity compositeEntity, IList<EntityRecommendation> entities, JObject entitiesAndMetadata, bool verbose)
+        {
+            var childrenEntites = new JObject();
+            var childrenEntitiesMetadata = new JObject();
+            if (verbose)
+            {
+                childrenEntites["$instance"] = new JObject();
+            }
+
+            // This is now implemented as O(n^2) search and can be reduced to O(2n) using a map as an optimization if n grows
+            var compositeEntityMetadata = entities.FirstOrDefault(e => e.Type == compositeEntity.ParentType && e.Entity == compositeEntity.Value);
+
+            // This is an error case and should not happen in theory
+            if (compositeEntityMetadata == null)
+                return entities;
+
+            if (verbose)
+            {
+                childrenEntitiesMetadata = GetEntityMetadata(compositeEntityMetadata);
+                childrenEntites["$instance"] = new JObject();
+            }
+
+
+            var filteredEntities = new List<EntityRecommendation>();
+            var coveredSet = new HashSet<int>();
+            foreach (var child in compositeEntity.Children)
+            {
+                for (var i = 0; i < entities.Count; i++)
+                {
+                    var entity = entities[i];
+                    if (!coveredSet.Contains(i) &&
+                        child.Type == entity.Type &&
+                        entity.StartIndex >= compositeEntityMetadata.StartIndex &&
+                        entity.EndIndex <= compositeEntityMetadata.EndIndex)
+                    {
+                        // Add to the set to ensure that we don't consider the same child entity more than once per composite
+                        coveredSet.Add(i);
+                        AddProperty(childrenEntites, GetNormalizedEntityType(entity), GetEntityValue(entity));
+
+                        if (verbose)
+                        {
+                            AddProperty((JObject)childrenEntites["$instance"], GetNormalizedEntityType(entity), GetEntityMetadata(entity));
+                        }
+                    }
+                }
+            }
+
+            // filter entities that were covered by this composite entity
+            for (var i = 0; i < entities.Count; i++)
```

Comment:
This seems weird. Is this code right? Why wouldn't this just be a standard for-each?

Reply:
As it sits, myself (and everyone looking at it) wonders, "Is it entities.Count, or entities.Count- 1?"

Reply:
Will refactor this